### PR TITLE
#1748 Upgrade Elyra Canvas to Node 20.x

### PR DESCRIPTION
**Pull request by srikant-ch5**
Fixes: #1748 
 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

File: canvas_modules/common-canvas/package.json [hunk 1]
```diff
@@ -15,7 +15,7 @@
   "license": "Apache-2.0",
   "sideEffects": false,
   "scripts": {
-    "build": "npx stylelint '**/*.scss' && grunt && rollup -c",
+    "build": "export NODE_OPTIONS=--openssl-legacy-provider && npx stylelint '**/*.scss' && grunt && rollup -c",
```
Comment (matthoward366) [inline canvas_modules/common-canvas/package.json:18]: Why is this needed here?

File: canvas_modules/harness/package.json [hunk 2]
```diff
@@ -4,15 +4,15 @@
   "private": true,
   "license": "Apache-2.0",
   "scripts": {
-    "start": "NODE_ENV=development node index.js",
-    "build": "npx stylelint 'src/**/*.scss' && grunt",
+    "start": "export NODE_OPTIONS=--openssl-legacy-provider && NODE_ENV=development node index.js",
```
Comment (matthoward366) [inline canvas_modules/harness/package.json:7]: @nmgokhale will this be fixed with your carbon 11 update and updating webpack.

File: canvas_modules/common-canvas/package.json (hunk 1 above)
Comment (srikant-ch5) [reply canvas_modules/common-canvas/package.json:18]: Hi Matt I have added this to resolve `Error: error:0308010C:digital envelope routines::unsupported` while upgrading from node 16, alternatively we can also add this in .npmrc file.

Comment (matthoward366) [reply canvas_modules/common-canvas/package.json:18]: That's a workaround.  What is needed to fix this correctly.

Comment (matthoward366) [reply canvas_modules/common-canvas/package.json:18]: This works fine for me in common-canvas
```
Running "postcss:dist" (postcss) task
Browserslist: caniuse-lite is outdated. Please run:
  npx update-browserslist-db@latest
  Why you should do it regularly: https://github.com/browserslist/update-db#readme
>> 1 processed stylesheet created.

Done.

./src/common-properties/index.js, ./src/common-properties/components/field-picker/index.js, ./src/common-properties/components/flexible-table/index.js, ./src/common-properties/controls/expression/languages/CLEM-hint.js, ./src/common-properties/controls/expression/languages/python-hint.js, ./src/context-menu/context-menu-wrapper.jsx, ./src/command-stack/command-stack.js, ./src/tooltip/tooltip.jsx, ./src/common-canvas/index.js, ./src/common-canvas/canvas-controller.js, ./src/index.js → ./dist, ./dist...
`isModuleDeclaration` has been deprecated, please migrate to `isImportOrExportDeclaration`
    at isModuleDeclaration (/Users/mhoward/github/mhoward/canvas/canvas_modules/common-canvas/node_modules/@babel/types/lib/validators/generated/index.js:2794:35)
    at PluginPass.Program (/Users/mhoward/github/mhoward/canvas/canvas_modules/common-canvas/node_modules/babel-plugin-lodash/lib/index.js:102:44)
(!) Circular dependencies
src/toolbar/toolbar-sub-menu.jsx -> src/toolbar/toolbar-sub-menu-item.jsx -> src/toolbar/toolbar-sub-menu.jsx
src/common-properties/controls/expression/expression.jsx -> src/common-properties/controls/expression/expression-builder/expression-builder.jsx -> src/common-properties/controls/expression/expression.jsx
src/common-properties/controls/control-factory.js -> src/common-properties/controls/structureeditor/index.js -> src/common-properties/controls/structureeditor/structureeditor.jsx -> src/common-properties/controls/control-factory.js
...and 15 more
created ./dist, ./dist in 11.5s
[mhoward@mbp ~/github/mhoward/canvas/canvas_modules/common-canvas (master)]$ node -v
v20.11.0
```

Comment (srikant-ch5) [reply canvas_modules/common-canvas/package.json:18]: I added this because it was failing during build https://github.com/elyra-ai/canvas/actions/runs/8229863100/job/22501894502#step:5:42 .

File: canvas_modules/harness/package.json (hunk 2 above)
Comment (nmgokhale) [reply canvas_modules/harness/package.json:7]: In my carbon 11 PR, these changes work. Didn't add `export NODE_OPTIONS=--openssl-legacy-provider &&` in any of the files. Test harness app runs OK. I didn't test all the components though -
Screenshot 2024-03-11 at 10 13 25 AM

Screenshot 2024-03-11 at 10 13 50 AM

**Review by matthoward366 — CHANGES_REQUESTED** [review]
Why so many changes. I think this is going to cause a lot of conflicts with what Neha is already doing.  My previous comment was only common-canvas.  I think in the test harness having the legacy parameter is fine for now.

I would revert this back to the previous comment with minor changes and let Neha deal with webpack update for carbon 11.

**Comment by srikant-ch5** [comment]
> Why so many changes. I think this is going to cause a lot of conflicts with what Neha is already doing. My previous comment was only common-canvas. I think in the test harness having the legacy parameter is fine for now.
> 
> I would revert this back to the previous comment with minor changes and let Neha deal with webpack update for carbon 11.

Hi Matt. The OpenSSL issue after node upgrade was due to webpack version so I was testing with webpack upgrade to see  whether PR build returns any error related to Node (because in my local npm run build is working fine but failing in PR). I have reverted webpack changes and also removed NODE_OPTIONS.

**Comment by srikant-ch5** [comment]
> For the test harness you can keep the `export NODE_OPTIONS=--openssl-legacy-provider`. Neha will remove this in her PR.

Hi Matt I have included above NODE_OPTIONS in test harness and updated PR. Thanks.

**Review by matthoward366 — APPROVED** [review]
